### PR TITLE
Fix overrided settings: normalize_function_names

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix overrided settings value of `normalize_function_names`.  

####  Bug description: 
Clickhouse cluster with 2 shards, each shard contains 2 replicas, and setting: `normalize_function_names` is set to true in each replica's config file.  
- shard01: r11, r12
- shard02: r21, r22 

1. Login r22 and query value of setting `normalize_function_names` from all replicas. Only value returned from local replica r22 is true. 
```
r22 :)
SELECT
    getMacro('replica') AS replica,
    max(value)
FROM clusterAllReplicas('preonline_hk5', system, settings)
WHERE name = 'normalize_function_names'
GROUP BY replica
FORMAT TSVWithNames

Query id: 8be6b7d0-35d3-49c5-9a30-c1ce0a8dc667

replica	max(value)
r11	0
r12	0
r21	0
r22	1

4 rows in set. Elapsed: 0.025 sec. Processed 2.26 thousand rows, 423.21 KB (91.97 thousand rows/s., 17.19 MB/s.)
```

2. Login r11 and query setting `normalize_function_names` of r11.  
```
r11 :) 
SELECT *
FROM system.settings
WHERE name = 'normalize_function_names'

Query id: 909c4192-1f06-454a-be83-5dc975036d53

Row 1:
──────
name:        normalize_function_names
value:       1
changed:     0
description: Normalize function names to their canonical names
min:         ᴺᵁᴸᴸ
max:         ᴺᵁᴸᴸ
readonly:    0
type:        Bool

1 rows in set. Elapsed: 0.004 sec. 
```

We can see that the value of setting `normalize_function_names` from remote replicas(r11, r12, r21) is not right. It should be `true`.  

#### Reason
In `TCPHandler::receiveQuery`, setting `normalize_function_names` is set to false in `query_context` if current query is secondary query. 
``` cpp
/// Disable function name normalization when it's a secondary query, because queries are either
    /// already normalized on initiator node, or not normalized and should remain unnormalized for
    /// compatibility.
    if (query_kind == ClientInfo::QueryKind::SECONDARY_QUERY)
    {
        query_context->setSetting("normalize_function_names", false);
    }
```

In `StorageSystemSettings::fillData`, all values of settings are extract from `query_context`. Which means `normalize_function_names` value queried from `system.settings` is false if current query is secondary query.   
``` cpp 

void StorageSystemSettings::fillData(MutableColumns & res_columns, ContextPtr context, const SelectQueryInfo &) const
{
    const Settings & settings = context->getSettingsRef();
    auto constraints_and_current_profiles = context->getSettingsConstraintsAndCurrentProfiles();
    const auto & constraints = constraints_and_current_profiles->constraints;
```

#### Solution 
In `StorageSystemSettings::fillData`, if current query is secondary query from initial CH node, get all settings from global context.

### CODE REVIEW

**Review by tavplubix — CHANGES_REQUESTED** [review]
This behavior is even more unexpected that implicitly changed `normalize_function_names` setting. Settings in global context may be completely different from settings for a user executing this query. Currently you get "incorrect" value for one setting, but after this change you will get incorrect values for all settings.

**Comment by tavplubix** [comment]
What we can try to do: normalize query if the setting is true and query is not secondary. It will not require to modify the setting for secondary queries.

**Comment by tavplubix** [comment]
(btw, we have other settings that can be changed during query execution, let's also check them)

**Comment by taiyang-li** [comment]
> What we can try to do: normalize query if the setting is true and query is not secondary. It will not require to modify the setting for secondary queries.

Good idea.  I will try it

File: src/Interpreters/TreeRewriter.cpp [hunk 1]
```diff
@@ -1363,7 +1363,10 @@ void TreeRewriter::normalize(
     MarkTableIdentifiersVisitor(identifiers_data).visit(query);
 
     /// Rewrite function names to their canonical ones.
-    if (settings.normalize_function_names)
+    /// Notice: function name normalization is disabled when it's a secondary query, because queries are either
+    /// already normalized on initiator node, or not normalized and should remain unnormalized for
+    /// compatibility.
+    if (context->getClientInfo().query_kind != ClientInfo::QueryKind::SECONDARY_QUERY && settings.normalize_function_names)
```
Comment (tavplubix) [inline src/Interpreters/TreeRewriter.cpp:1369]: It's not the only place where this setting is used

Comment (taiyang-li) [reply src/Interpreters/TreeRewriter.cpp:1369]: I'm not sure if we should skip the whole `TreeRewriter::normalize` if query is secondary. @tavplubix  Do you think it is ok ?

Comment (tavplubix) [reply src/Interpreters/TreeRewriter.cpp:1369]: Why do we need to skip the whole `TreeRewriter::normalize`?

Comment (taiyang-li) [reply src/Interpreters/TreeRewriter.cpp:1369]: > It's not the only place where this setting is used

Sorry, I misunderstood. Get it now.

Comment (taiyang-li) [reply src/Interpreters/TreeRewriter.cpp:1373]: > It's not the only place where this setting is used

@tavplubix  done.

**Comment by tavplubix** [comment]
Fast test (actions) — Cannot build ClickHouse

**Comment by taiyang-li** [comment]
> 

fixed.

**Comment by tavplubix** [comment]
It's still broken

**Comment by taiyang-li** [comment]
> It's still broken

Commit again, let's wait

**Pull request merged**